simba/optimizer_util.py

```diff
@@ -294,39 +294,53 @@ def get_rotation_soc_util(rot_id, schedule, scenario, soc_data: dict = None):
     return scenario.vehicle_socs[rot.vehicle_id], rot_start_idx, rot_end_idx
 
 
-def get_delta_soc(soc_over_time_curve, soc, time_delta, optimizer: 'StationOptimizer'):
+def get_delta_soc(soc_over_time_curve, soc, duration_min: float):
     """ Return expected SoC lift for a given SoC charging time series, start_soc and time_delta.
 
     :param soc_over_time_curve: Data with columns: time, soc and n rows
     :type soc_over_time_curve: np.array() with shape(n, 2)
     :param soc: start socs
     :type soc: float
-    :param time_delta: time of charging
-    :type time_delta: float
-    :param optimizer: optimizer object
-    :type optimizer: simba.station_optimizer.StationOptimizer
+    :param duration_min: duration of charging in minutes
+    :type duration_min: float
     :return: positive delta of the soc
     :rtype: float
     """
     # units for time_delta and time_curve are assumed to be the same, e.g. minutes
     # first element which is bigger than current soc
-    if time_delta == 0:
+
+    if duration_min == 0:
         return 0
-    soc = max(min(optimizer.args.desired_soc_opps, soc), 0)
+    negative_soc = 0
+    if soc < 0:
+        # keep track of the negative soc to add it later
+        negative_soc = -soc
+        try:
+            gradient = soc_over_time_curve[1, 1] - soc_over_time_curve[0, 1]
```

Comment:
you are assuming the soc over time curve to be a single linear gradient? Values are taken from the initial state and the next timestep?

Reply:
the gradient at the lowest soc of the charging curve to the second lowest soc of the charging curve is extrapolated to negative values.
e.g. an initial charge curve of
[0,100],[0.8,100],[0.9,20]
would be numerically solved over time. e.g
time = 0s --> soc[0]  = 0
time  = 1s --> soc[1] = soc[0] + 100kW*1s/(battery capacity)
time = 2s --> soc[2] = soc[1] + 100kW*1s/(battery capacity)
the gradient between time=0s and 1s is used for charging while the soc is negative. AFAIK thats the same method SpiceEV uses.